File: chap1/sec1.3/PlotTransport.m
function PlotTransport(u,x,t,u0);
% PLOTTRANSPORT plots an numerical and exact transport solution
%   PlotTransport(u,x,t,u0); plots a numerical solution in u on the
%   space grid in x and time grid in t (colums of u are space),
%   representing an approximate solution for an advection equation,
%   and also the exact solution from the initial condition is plotted
%   if given by the initial condition function u0.

ma=max(max(u));
mi=min(min(u));
a=x(1);
b=x(length(x));
xx=(a:(b-a)/500:b)';
for n=1:size(u,2)
  if nargin==4
    plot(x,u(:,n),'o',xx,feval(u0,xx-t(n)),'-','LineWidth',2,'Markersize',10);
  else      
    plot(x,u(:,n),'o')
  end  
  xlabel('x');ylabel('u');
  t(n)
  pause
end;

File: chap1/sec1.3/PlotTransport.py
import numpy as np
import matplotlib.pyplot as plt

def plotTransport(u, x, t, u0=None):
    """
    Plots the numerical solution of the transport equation and optionally
    compares it with the exact solution derived from the initial condition.

    Parameters
    ----------
    u : ndarray
        Solution array of shape (J+1, N+1), where each column corresponds to time.
    x : ndarray
        Spatial grid points.
    t : ndarray
        Time grid points.
    u0 : function, optional
        Initial condition function u0(x). If provided, the exact solution is plotted for comparison.
    """
    ma = np.max(u)
    mi = np.min(u)
    a = x[0]
    b = x[-1]
    
    # High-resolution spatial grid for exact solution
    xx = np.linspace(a, b, 500)
    
    for n in range(u.shape[1]):  # Iterate over time steps
        plt.clf()
        if u0 is not None:
            exact_solution = u0(xx - t[n])
            plt.plot(x, u[:, n], 'o', label='Numerical Solution', markersize=8)
            plt.plot(xx, exact_solution, '-', label='Exact Solution', linewidth=2)
        else:
            plt.plot(x, u[:, n], 'o', label='Numerical Solution', markersize=8)
        
        plt.xlabel('x')
        plt.ylabel('u')
        plt.ylim([mi, ma])
        plt.title(f'Time t = {t[n]:.2f}')
        plt.legend()
      
